webapp/channels/src/utils/desktopAPI.ts

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import semver from 'semver';
+
+import {isDesktopApp} from 'utils/user_agent';
+
+// TODO: Temporary typings
+import type {DesktopAPI} from '../../../../../desktop/src/types/api/main';
+declare global {
+    interface Window {
+        desktopAPI?: DesktopAPI;
+    }
+}
+
+class DesktopApp {
+    private name?: string;
+    private version?: string | null;
+    private dev?: boolean;
+
+    /**
+     * @deprecated
+     */
+    private postMessageListeners?: Map<string, Set<(message: unknown) => void>>;
+
+    constructor() {
+        // Check the user agent string first
+        if (!isDesktopApp()) {
+            return;
+        }
+
+        this.getDesktopAppInfo().then(({name, version}) => {
+            this.name = name;
+            this.version = semver.valid(semver.coerce(version));
+
+            // Legacy Desktop App version, used by some plugins
+            if (!window.desktop) {
+                window.desktop = {};
+            }
+            window.desktop.version = semver.valid(semver.coerce(version));
+        });
+        window.desktopAPI?.isDev?.().then((isDev) => {
+            this.dev = isDev;
+        });
+
+        // Legacy code - to be removed
+        this.postMessageListeners = new Map();
```

Comment:
To avoid having the rest of the app care about whether we use `window.postMessage` or the new API, we wrap a class here containing a set of listeners we register to be called when the `message` event comes down.

All of this will go away once we remove `window.postMessage` permanently.